presto-orc/src/test/java/com/facebook/presto/orc/BenchmarkMapFlatWriter.java

```diff
@@ -123,64 +129,125 @@ private void doWrite(BenchmarkData data)
+            Type keyType = type.getKeyType();
+            Type valueType = type.getValueType();
 
-            this.type = mapType(keyType, valueType);
             this.temporaryDirectory = createTempDir();
             this.orcFile = new File(temporaryDirectory, randomUUID().toString());
+            this.orcFile.deleteOnExit();
 
             ImmutableList.Builder<Block> blocks = ImmutableList.builder();
 
-            List<Long> keys = new ArrayList<>();
+            List<Long> keys = new ArrayList<>(distinctKeyCount);
             for (int i = 0; i < distinctKeyCount; i++) {
                 keys.add((long) i);
             }
+            Collections.shuffle(keys, random);
+
+            List<Integer> weights = Arrays.stream(frequencyHistogram.split(","))
+                    .map(String::trim)
+                    .map(Integer::valueOf)
+                    .collect(Collectors.toList());
+            assertEquals(weights.size(), 11, "Number of weights is expected to be 11");
+
+            double weightSum = weights.stream().mapToInt(Integer::intValue).sum();
+            double[] keyFrequencies = new double[distinctKeyCount];
+            int offset = 0;
+            for (int i = 0; i < weights.size(); i++) {
+                double frequency = 0.1 * (i + 1);
+                int keyCount = (int) (distinctKeyCount * weights.get(i) / weightSum);
+                for (int j = 0; j < keyCount; j++) {
+                    keyFrequencies[j + offset] = frequency;
+                }
+                offset += keyCount;
+            }
 
             for (int i = 0; i < blockCount; i++) {
-                MapBlockBuilder mapBlockBuilder = (MapBlockBuilder) type.createBlockBuilder(null, positionsPerBlock);
-                for (int j = 0; j < positionsPerBlock; j++) {
-                    Collections.shuffle(keys, random);
-                    int entryKeyCount = random.nextInt(maxKeysPerEntry);
+                MapBlockBuilder mapBlockBuilder = (MapBlockBuilder) this.type.createBlockBuilder(null, positionsPerBlock);
+                BlockBuilder mapKeyBuilder = mapBlockBuilder.getKeyBlockBuilder();
+                BlockBuilder mapValueBuilder = mapBlockBuilder.getValueBlockBuilder();
 
+                for (int j = 0; j < positionsPerBlock; j++) {
                     mapBlockBuilder.beginDirectEntry();
-                    BlockBuilder keyBuilder = mapBlockBuilder.getKeyBlockBuilder();
-                    BlockBuilder valueBuilder = mapBlockBuilder.getValueBlockBuilder();
 
-                    for (int k = 0; k < entryKeyCount; k++) {
-                        keyType.writeLong(keyBuilder, keys.get(k));
-                        keyType.writeLong(valueBuilder, keys.get(k) * 2L + 1);
+                    int entryKeyCount = 0;
+
+                    // add keys
+                    for (int k = 0; k < distinctKeyCount; k++) {
+                        if (random.nextDouble() < keyFrequencies[k]) {
```

Comment:
nit: recommend generate random.nextDoulbe once outside the loop and use it always. This is generated again and again and it is hard to reason about.

Reply:
In my opinion generating a new threshold for each key is better because if we do it once per map then we will target all keys in the probability bucket. I'd rather give each key a bit more of individual attention.